truffleruby/src/main/c/cext/ruby.c

```diff
@@ -875,10 +875,12 @@ VALUE rb_str_new(const char *string, long length) {
 }
 
 VALUE rb_str_new_cstr(const char *string) {
+  int len = strlen(string);
   if (truffle_is_truffle_object((VALUE) string)) {
-    return (VALUE) truffle_invoke(RUBY_CEXT, "to_ruby_string", string);
+    VALUE ruby_string = (VALUE) truffle_invoke(RUBY_CEXT, "to_ruby_string", string);
+    return (VALUE) truffle_invoke(ruby_string, "[]", 0, len);
   } else {
-    return (VALUE) truffle_invoke(RUBY_CEXT, "rb_str_new_cstr", truffle_read_string(string));
+    return (VALUE) truffle_invoke(RUBY_CEXT, "rb_str_new_cstr", truffle_read_n_string(string, len));
```

Comment:
This traverses twice the C string, could we move the `strlen` call to only the TruffleObject branch?